x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/AuthHeadersExtractor.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ml.utils;
+
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationField;
+import org.elasticsearch.xpack.core.security.authc.support.SecondaryAuthentication;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public final class AuthHeadersExtractor {
+
+    private AuthHeadersExtractor() {}
+
+    public static Map<String, String> extractAuthHeadersAndPreferSecondaryAuth(ThreadContext context) {
+        Map<String, String> threadHeaders = context.getHeaders();
+        if (threadHeaders == null || threadHeaders.isEmpty()) {
+            return new HashMap<>();
+        }
+
+        Map<String, String> headers = threadHeaders.entrySet().stream()
+            .filter(e -> ClientHelper.SECURITY_HEADER_FILTERS.contains(e.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        String secondaryAuth = headers.get(SecondaryAuthentication.THREAD_CTX_KEY);
+        if (secondaryAuth != null) {
+            headers.put(AuthenticationField.AUTHENTICATION_KEY, secondaryAuth);
+        }
+        return headers;
```

Comment:
@tvernum We are storing `_xpack_security_secondary_authc` as the primary auth for some of our tasks. This way we "prefer" the secondary auth if it was provided.

Is simply replacing `_xpack_security_authentication` with the contents of `_xpack_security_secondary_authc` acceptable? Or are there technical limitations/details that I am missing?

Reply:
Following the test you did yesterday this looks like it works as intended now.

Reply:
This isn't how I expected you would do it...

I had presumed that you would _activate_ the secondary authentication so that it became the only authentication and then just do what you've always done.

Something like:

```
Runnable body = () -> {
   final Map<String, String> headers = threadPool.getThreadContext().getHeaders();   
   datafeedConfigProvider.updateDatefeedConfig(
            request.getUpdate().getId(),
            request.getUpdate(),
            headers,
            jobConfigProvider::validateDatafeedJob,
            ActionListener.wrap(
                updatedConfig -> listener.onResponse(new PutDatafeedAction.Response(updatedConfig)),
                listener::onFailure));
};
SecondaryAuthentication secondaryAuth = SecondaryAuthentication.readFromContext(securityContext);
if (secondaryAuth != null) {
 body = secondaryAuth.wrap(body);
}
body.run();
```

We could definitely, reduce some of that boilerplate and move it into `SecondaryAuthentication` if it would help, but the above is how I expected secondary auth to be used.

Reply:
@tvernum we need the headers to be respected by `_search` and apply the roles that are stored in that context. 

Do the other APIs respect secondary auth headers?

Another pain point is BWC. As a 'new' datafeed created with secondary auth headers could be used on an 'old' node. Then security would know nothing of the secondary auth concept and the datafeed would fail.

Reply:
Ah, I think I understand. `secondaryAuth.wrap` changes the stored headers so that the secondary auth header becomes the `Authorization` header. 

Seems more idiomatic than iterating a map :). Let experiment and see if I can get it to work this way.